server/src/main/java/org/elasticsearch/rest/action/search/RestSearchAction.java

```diff
@@ -145,54 +148,75 @@ public RestChannelConsumer prepareRequest(final RestRequest request, final NodeC
      * @param setSize how the size url parameter is handled. {@code udpate_by_query} and regular search differ here.
      */
     public static void parseSearchRequest(
         SearchRequest searchRequest,
-        RestRequest request,
+        RestApiVersion restApiVersion,
+        RequestParams params,
         XContentParser requestContentParser,
-        NamedWriteableRegistry namedWriteableRegistry,
         Predicate<NodeFeature> clusterSupportsFeature,
         IntConsumer setSize
     ) throws IOException {
-        parseSearchRequest(searchRequest, request, requestContentParser, namedWriteableRegistry, clusterSupportsFeature, setSize, null);
+        parseSearchRequest(
+            searchRequest,
+            restApiVersion,
+            params,
+            requestContentParser,
+            clusterSupportsFeature,
+            setSize,
+            null,
+            SearchRequestType.SEARCH
+        );
+    }
+
+    public enum SearchRequestType {
+        SEARCH,
+        MULTI_SEARCH
     }
 
     /**
      * Parses the rest request on top of the SearchRequest, preserving values that are not overridden by the rest request.
      *
      * @param searchRequest the search request that will hold what gets parsed
-     * @param request the rest request to read from
+     * @param restApiVersion REST API version for the request
+     * @param params request parameters
      * @param requestContentParser body of the request to read. This method does not attempt to read the body from the {@code request}
      *        parameter, will be null when there is no request body to parse
-     * @param namedWriteableRegistry the registry of named writeables
       @param clusterSupportsFeature used to check if certain features are available in this cluster
      * @param setSize how the size url parameter is handled. {@code udpate_by_query} and regular search differ here.
      * @param searchUsageHolder the holder of search usage stats
+     * @param searchRequestType the type of search request
      */
     public static void parseSearchRequest(
         SearchRequest searchRequest,
-        RestRequest request,
+        RestApiVersion restApiVersion,
+        RequestParams params,
         @Nullable XContentParser requestContentParser,
-        NamedWriteableRegistry namedWriteableRegistry,
         Predicate<NodeFeature> clusterSupportsFeature,
         IntConsumer setSize,
-        @Nullable SearchUsageHolder searchUsageHolder
+        @Nullable SearchUsageHolder searchUsageHolder,
+        SearchRequestType searchRequestType
+
     ) throws IOException {
-        if (request.getRestApiVersion() == RestApiVersion.V_7 && request.hasParam("type")) {
-            request.param("type");
-            deprecationLogger.compatibleCritical("search_with_types", TYPES_DEPRECATION_MESSAGE);
+        if (restApiVersion == RestApiVersion.V_7 && params.hasParam("type")) {
```

Comment:
This is ugly but necessary - we need to understand what search type we have so we return the appropriate deprecation warning.